Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

```diff
@@ -13,9 +14,146 @@ MacroscopicProperties::ReadParameters ()
-    pp.get("sigma", m_sigma);
-    pp.get("mu", m_mu);
-    pp.get("epsilon", m_epsilon);
+    pp.get("sigma_init_style", m_sigma_s);
+    // constant initialization
+    if (m_sigma_s == "constant") pp.get("sigma", m_sigma);
+    // initialization of sigma (conductivity) with parser
+    if (m_sigma_s == "parse_sigma_function") {
+        Store_parserString(pp, "sigma_function(x,y,z)", m_str_sigma_function);
+        m_sigma_parser.reset(new ParserWrapper<3>(
+                                 makeParser(m_str_sigma_function,{"x","y","z"}) ) );
+    }
+
+    pp.get("epsilon_init_style", m_epsilon_s);
+    if (m_epsilon_s == "constant") pp.get("epsilon", m_epsilon);
+    // initialization of epsilon (permittivity) with parser
+    if (m_epsilon_s == "parse_epsilon_function") {
+        Store_parserString(pp, "epsilon_function(x,y,z)", m_str_epsilon_function);
+        m_epsilon_parser.reset(new ParserWrapper<3>(
+                                 makeParser(m_str_epsilon_function,{"x","y","z"}) ) );
+    }
+
+    pp.get("mu_init_style", m_mu_s);
+    if (m_mu_s == "constant") pp.get("mu", m_mu);
+    // initialization of mu (permeability) with parser
+    if (m_mu_s == "parse_mu_function") {
+        Store_parserString(pp, "mu_function(x,y,z)", m_str_mu_function);
+        m_mu_parser.reset(new ParserWrapper<3>(
+                                 makeParser(m_str_mu_function,{"x","y","z"}) ) );
+    }
+
+}
+
+void
+MacroscopicProperties::InitData ()
+{
+    amrex::Print() << "we are in init data of macro \n";
+    auto & warpx = WarpX::GetInstance();
+
+    // Get BoxArray and DistributionMap of warpx instant.
+    int lev = 0;
+    BoxArray ba = warpx.boxArray(lev);
+    DistributionMapping dmap = warpx.DistributionMap(lev);
+    int ng = 3;
+    // Define material property multifabs using ba and dmap from WarpX instance
+    // sigma is cell-centered MultiFab
+    m_sigma_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+    // epsilon is cell-centered MultiFab
+    m_eps_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+    // mu is cell-centered MultiFab
+    m_mu_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+    // Initialize sigma
+    if (m_sigma_s == "constant") {
+
+        m_sigma_mf->setVal(m_sigma);
+
+    } else if (m_sigma_s == "parse_sigma_function") {
+
+        InitializeMacroMultiFabUsingParser(m_sigma_mf.get(), m_sigma_parser.get(), lev);
+    }
+    // Initialize epsilon
+    if (m_epsilon_s == "constant") {
+
+        m_eps_mf->setVal(m_epsilon);
+
+    } else if (m_epsilon_s == "parse_epsilon_function") {
+
+        InitializeMacroMultiFabUsingParser(m_eps_mf.get(), m_epsilon_parser.get(), lev);
+
+    }
+    // Initialize mu
+    if (m_mu_s == "constant") {
+
+        m_mu_mf->setVal(m_mu);
+
+    } else if (m_mu_s == "parse_mu_function") {
+
+        InitializeMacroMultiFabUsingParser(m_mu_mf.get(), m_mu_parser.get(), lev);
+
+    }
+
+
+    sigma_IndexType.resize( 3 );
```

Comment:
Ah Thank you for catching this. This will be 3, because I am using the function CoarsenIO::Interp(), defined in Source/Utils/CoarsenIO.H that works with GPU arrays of size 3.
But I realized after you brought this up, that I initialize their values only for AMREX_SPACEDIM elements.
So now I have added a commit to give the last element 0 value if 2D.

Alll these arrays will be removed in the "directly_apply_parser-PR" -- next in my to-do.